File: ios/Runner/Ble/BleMethodChannel.swift
import Foundation
import Flutter
import CoreBluetooth

/// BLE Peripheral用MethodChannel
/// FlutterとネイティブのBluetooth Peripheral機能を連携
class BleMethodChannel: NSObject {
    private static let methodChannelName = "com.pixeldiary/ble_peripheral"
    private static let eventChannelName = "com.pixeldiary/ble_peripheral/events"

    private let peripheralManager: BlePeripheralManager
    private var eventSink: FlutterEventSink?

    init(binaryMessenger: FlutterBinaryMessenger) {
        self.peripheralManager = BlePeripheralManager()
        super.init()

        // MethodChannel設定
        let methodChannel = FlutterMethodChannel(
            name: BleMethodChannel.methodChannelName,
            binaryMessenger: binaryMessenger
        )
        methodChannel.setMethodCallHandler(handleMethodCall)

        // EventChannel設定
        let eventChannel = FlutterEventChannel(
            name: BleMethodChannel.eventChannelName,
            binaryMessenger: binaryMessenger
        )
        eventChannel.setStreamHandler(self)

        // コールバック設定
        setupCallbacks()
    }

    // MARK: - Method Call Handler

    private func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startAdvertising":
            startAdvertising(call, result: result)
        case "stopAdvertising":
            stopAdvertising(result: result)
        case "isAdvertising":
            isAdvertising(result: result)
        case "getConnectedDeviceCount":
            getConnectedDeviceCount(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Implementations

    /// アドバタイズ開始
    private func startAdvertising(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let nickname = args["nickname"] as? String,
              let artData = args["artData"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "nickname and artData are required",
                details: nil
            ))
            return
        }

        print("[BleMethodChannel] Starting advertising with nickname: \(nickname)")

        let success = peripheralManager.startAdvertising(nickname: nickname, artData: artData)

        if success {
            result(true)
        } else {
            result(FlutterError(
                code: "START_FAILED",
                message: "Failed to start advertising",
                details: nil
            ))
        }
    }

    /// アドバタイズ停止
    private func stopAdvertising(result: @escaping FlutterResult) {
        print("[BleMethodChannel] Stopping advertising")
        peripheralManager.stopAdvertising()
        result(true)
    }

    /// アドバタイズ状態確認
    private func isAdvertising(result: @escaping FlutterResult) {
        let advertising = peripheralManager.getIsAdvertising()
        result(advertising)
    }

    /// 接続デバイス数取得
    private func getConnectedDeviceCount(result: @escaping FlutterResult) {
        let count = peripheralManager.getConnectedDeviceCount()
        result(count)
    }

    // MARK: - Callbacks

    private func setupCallbacks() {
        // データ受信時
        peripheralManager.onDataReceived = { [weak self] data, central in
            print("[BleMethodChannel] Data received from central")
            self?.eventSink?([
                "type": "dataReceived",
                "data": data,
                "deviceAddress": central.identifier.uuidString,
                "deviceName": "Unknown"
            ])
        }

        // デバイス接続時
        peripheralManager.onDeviceConnected = { [weak self] central in
            print("[BleMethodChannel] Device connected")
            self?.eventSink?([
                "type": "deviceConnected",
                "deviceAddress": central.identifier.uuidString,
                "deviceName": "Unknown"
            ])
        }

        // デバイス切断時
        peripheralManager.onDeviceDisconnected = { [weak self] central in
            print("[BleMethodChannel] Device disconnected")
            self?.eventSink?([
                "type": "deviceDisconnected",
                "deviceAddress": central.identifier.uuidString,
                "deviceName": "Unknown"
            ])
        }

        // エラー時
        peripheralManager.onError = { [weak self] message in
            print("[BleMethodChannel] Peripheral error: \(message)")
            self?.eventSink?(FlutterError(
                code: "BLE_ERROR",
                message: message,
                details: nil
            ))
        }

        // アドバタイズ開始時
        peripheralManager.onAdvertisingStarted = { [weak self] in
            print("[BleMethodChannel] Advertising started")
            self?.eventSink?([
                "type": "advertisingStarted"
            ])
        }

        // アドバタイズ停止時
        peripheralManager.onAdvertisingStopped = { [weak self] in
            print("[BleMethodChannel] Advertising stopped")
            self?.eventSink?([
                "type": "advertisingStopped"
            ])
        }
    }

    // MARK: - Cleanup

    func cleanup() {
        peripheralManager.cleanup()
        eventSink = nil
    }
}

// MARK: - FlutterStreamHandler

extension BleMethodChannel: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        print("[BleMethodChannel] Event channel listener attached")
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        print("[BleMethodChannel] Event channel listener detached")
        return nil
    }
}

File: android/app/src/main/kotlin/com/example/pixeldiary/ble/BleMethodChannel.kt
package com.example.pixeldiary.ble

import android.content.Context
import android.util.Log
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/// BLE Peripheral用MethodChannel
/// FlutterとネイティブのBluetooth Peripheral機能を連携
class BleMethodChannel(
    private val context: Context,
    methodChannel: MethodChannel,
    eventChannel: EventChannel
) : MethodChannel.MethodCallHandler {

    companion object {
        private const val TAG = "BleMethodChannel"
    }

    private val peripheralManager = BlePeripheralManager(context)
    private var eventSink: EventChannel.EventSink? = null

    init {
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
                Log.d(TAG, "Event channel listener attached")
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
                Log.d(TAG, "Event channel listener detached")
            }
        })

        // BlePeripheralManagerのコールバック設定
        setupCallbacks()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startAdvertising" -> startAdvertising(call, result)
            "stopAdvertising" -> stopAdvertising(result)
            "isAdvertising" -> isAdvertising(result)
            "getConnectedDeviceCount" -> getConnectedDeviceCount(result)
            else -> result.notImplemented()
        }
    }

    /// アドバタイズ開始
    private fun startAdvertising(call: MethodCall, result: MethodChannel.Result) {
        try {
            val nickname = call.argument<String>("nickname")
            val artData = call.argument<String>("artData")

            if (nickname == null || artData == null) {
                result.error("INVALID_ARGUMENTS", "nickname and artData are required", null)
                return
            }

            Log.d(TAG, "Starting advertising with nickname: $nickname")

            val success = peripheralManager.startAdvertising(nickname, artData)

            if (success) {
                result.success(true)
            } else {
                result.error("START_FAILED", "Failed to start advertising", null)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error starting advertising", e)
            result.error("EXCEPTION", e.message, null)
        }
    }

    /// アドバタイズ停止
    private fun stopAdvertising(result: MethodChannel.Result) {
        try {
            Log.d(TAG, "Stopping advertising")
            peripheralManager.stopAdvertising()
            result.success(true)
        } catch (e: Exception) {
            Log.e(TAG, "Error stopping advertising", e)
            result.error("EXCEPTION", e.message, null)
        }
    }

    /// アドバタイズ状態確認
    private fun isAdvertising(result: MethodChannel.Result) {
        try {
            val advertising = peripheralManager.isAdvertising()
            result.success(advertising)
        } catch (e: Exception) {
            Log.e(TAG, "Error checking advertising status", e)
            result.error("EXCEPTION", e.message, null)
        }
    }

    /// 接続デバイス数取得
    private fun getConnectedDeviceCount(result: MethodChannel.Result) {
        try {
            val count = peripheralManager.getConnectedDeviceCount()
            result.success(count)
        } catch (e: Exception) {
            Log.e(TAG, "Error getting connected device count", e)
            result.error("EXCEPTION", e.message, null)
        }
    }

    /// コールバック設定
    private fun setupCallbacks() {
        // データ受信時
        peripheralManager.onDataReceived = { data, device ->
            Log.d(TAG, "Data received from ${device.address}")
            eventSink?.success(
                mapOf(
                    "type" to "dataReceived",
                    "data" to data,
                    "deviceAddress" to device.address,
                    "deviceName" to (device.name ?: "Unknown")
                )
            )
        }

        // デバイス接続時
        peripheralManager.onDeviceConnected = { device ->
            Log.d(TAG, "Device connected: ${device.address}")
            eventSink?.success(
                mapOf(
                    "type" to "deviceConnected",
                    "deviceAddress" to device.address,
                    "deviceName" to (device.name ?: "Unknown")
                )
            )
        }

        // デバイス切断時
        peripheralManager.onDeviceDisconnected = { device ->
            Log.d(TAG, "Device disconnected: ${device.address}")
            eventSink?.success(
                mapOf(
                    "type" to "deviceDisconnected",
                    "deviceAddress" to device.address,
                    "deviceName" to (device.name ?: "Unknown")
                )
            )
        }

        // エラー時
        peripheralManager.onError = { message ->
            Log.e(TAG, "Peripheral error: $message")
            eventSink?.error("BLE_ERROR", message, null)
        }
    }

    /// クリーンアップ
    fun cleanup() {
        peripheralManager.cleanup()
        eventSink = null
    }
}
